server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java

```diff
@@ -390,14 +395,27 @@ public PipelinedRequest resolvePipeline(SearchRequest searchRequest) {
             if (searchRequest.pipeline() != null) {
                 // Named pipeline specified for the request
                 pipelineId = searchRequest.pipeline();
-            } else if (state != null && searchRequest.indices() != null && searchRequest.indices().length == 1) {
-                // Check for index default pipeline
-                IndexMetadata indexMetadata = state.metadata().index(searchRequest.indices()[0]);
-                if (indexMetadata != null) {
-                    Settings indexSettings = indexMetadata.getSettings();
-                    if (IndexSettings.DEFAULT_SEARCH_PIPELINE.exists(indexSettings)) {
-                        pipelineId = IndexSettings.DEFAULT_SEARCH_PIPELINE.get(indexSettings);
+            } else if (state != null && searchRequest.indices() != null && searchRequest.indices().length != 0) {
+                try {
+                    // Check for index default pipeline
+                    Index[] concreteIndices = indexNameExpressionResolver.concreteIndices(state, searchRequest);
+                    for (Index index : concreteIndices) {
+                        IndexMetadata indexMetadata = state.metadata().index(index);
+                        if (indexMetadata != null) {
+                            Settings indexSettings = indexMetadata.getSettings();
+                            if (IndexSettings.DEFAULT_SEARCH_PIPELINE.exists(indexSettings)) {
+                                String currentPipelineId = IndexSettings.DEFAULT_SEARCH_PIPELINE.get(indexSettings);
+                                if (NOOP_PIPELINE_ID.equals(pipelineId)) {
+                                    pipelineId = currentPipelineId;
+                                } else if (!pipelineId.equals(currentPipelineId)) {
+                                    pipelineId = NOOP_PIPELINE_ID;
+                                    break;
+                                }
+                            }
+                        }
                     }
+                } catch (IndexNotFoundException e) {
+                    logger.debug("Default pipeline not applied for {}", (Object) searchRequest.indices());
                 }
```

Comment:
I don't think we can just swallow the exception without changing any state.

In particular, suppose the request targets indices A, B, and C, where A and C have default (but different) pipelines and B doesn't exist. We would end up applying A's pipeline to C (assuming `ignore_unavailable` is `true`, so we skip B).

Or is the exception thrown by `concreteIndices`? If that's the case we might do the wrong thing when `ignore_unavailable` is `true` -- in the example above, if A and C use the same pipeline and B doesn't exist, I think we *would* want to apply the pipeline (since we're just ignoring B).

Reply:
Yes the exception is thrown from `concreteIndices` and it fails the below test if we don't have a try/catch here

```
REPRODUCE WITH: ./gradlew ':qa:ccs-unavailable-clusters:integTest' --tests "org.opensearch.search.CrossClusterSearchUnavailableClusterIT.testSearchSkipUnavailable" -Dtests.seed=D240A247ABFCD6B7 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=de-LU -Dtests.timezone=Pacific/Wake -Druntime.java=17
  2> [remote1:index] OpenSearchStatusException[OpenSearch exception [type=index_not_found_exception, reason=no such index [remote1:index]]]
        at __randomizedtesting.SeedInfo.seed([D240A247ABFCD6B7:8532D8AF58B6D5E]:0)
        at app//org.opensearch.rest.BytesRestResponse.errorFromXContent(BytesRestResponse.java:210)
        at app//org.opensearch.client.RestHighLevelClient.parseEntity(RestHighLevelClient.java:2235)
        at app//org.opensearch.client.RestHighLevelClient.parseResponseException(RestHighLevelClient.java:2212)
        at app//org.opensearch.client.RestHighLevelClient.internalPerformRequest(RestHighLevelClient.java:1931)
        at app//org.opensearch.client.RestHighLevelClient.performRequest(RestHighLevelClient.java:1884)
        at app//org.opensearch.client.RestHighLevelClient.performRequestAndParseEntity(RestHighLevelClient.java:1852)
        at app//org.opensearch.client.RestHighLevelClient.search(RestHighLevelClient.java:1095)
        at app//org.opensearch.search.CrossClusterSearchUnavailableClusterIT.testSearchSkipUnavailable(CrossClusterSearchUnavailableClusterIT.java:177)
  2> NOTE: leaving temporary files on disk at: /home/ubuntu/OpenSearch/qa/ccs-unavailable-clusters/build/testrun/integTest/temp/org.opensearch.search.CrossClusterSearchUnavailableClusterIT_D240A247ABFCD6B7-001
```

Reply:
I think the fundamental problem here was that I misunderstood the behavior of `concreteIndices`.

Since it takes the `SearchRequest` as input and the `SearchRequest` has the `indicesOptions` property, then `concreteIndices` will not throw an exception if `ignore_unavailable` is `true` -- it will just silently skip B without throwing an exception. So, the behavior I described above should happen "for free".

Part of me is still a little nervous about swallowing the exception, but it will still get thrown later, when we try to determine what shards to target. So, it should be good. :+1: